File: BOJ/[BOJ 1074] Z.py
'''
편의상 

1사분면 | 2사분면
ㅡㅡㅡㅡㅡㅡㅡㅡㅡ
3사분면 | 4사분면

으로 가정하고 풀이 진행

'''

N, r, c = map(int, input().split())

def recur(N, r, c):
    global total
    
    if N == 0:
        return
    
    m = 2**(N-1)
    if r < m and c < m:     # 1사분면
        pass
    elif r < m and c >= m:   # 2사분면
        total += (m**2)
    elif r >= m and c < m:   # 3사분면
        total += (m**2 * 2)
    elif r >= m and c >= m:   # 4사분면
        total += (m**2 * 3)
    recur(N-1, r%m, c%m)

total = 0
recur(N, r, c)
print(total)

File: BOJ/[BOJ 1074] Z.js
const fs = require("fs");
const input = fs
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split(" ")
  .map(Number);

const [N, r, c] = input;

let total = 0;
const recur = (n, r, c) => {
  const k = 2 ** (n - 1);
  if (n === 0) {
    return;
  }
  if (r < k && c < k) {
  } else if (r < k && c >= k) {
    total += k ** 2;
  } else if (r >= k && c < k) {
    total += k ** 2 * 2;
  } else if (r >= k && c >= k) {
    total += k ** 2 * 3;
  }
  recur(n - 1, r % k, c % k);
};
recur(N, r, c);
console.log(total);
